include/circt-c/Firtool/Firtool.h

```diff
@@ -62,83 +68,126 @@ typedef enum FirtoolRandomKind {
-DECLARE_FIRTOOL_OPTION(LowerMemories, bool);
-DECLARE_FIRTOOL_OPTION(BlackBoxRootPath, MlirStringRef);
-DECLARE_FIRTOOL_OPTION(ReplSeqMem, bool);
-DECLARE_FIRTOOL_OPTION(ReplSeqMemFile, MlirStringRef);
-DECLARE_FIRTOOL_OPTION(ExtractTestCode, bool);
-DECLARE_FIRTOOL_OPTION(IgnoreReadEnableMem, bool);
-DECLARE_FIRTOOL_OPTION(DisableRandom, FirtoolRandomKind);
-DECLARE_FIRTOOL_OPTION(OutputAnnotationFilename, MlirStringRef);
-DECLARE_FIRTOOL_OPTION(EnableAnnotationWarning, bool);
-DECLARE_FIRTOOL_OPTION(AddMuxPragmas, bool);
-DECLARE_FIRTOOL_OPTION(EmitChiselAssertsAsSVA, bool);
-DECLARE_FIRTOOL_OPTION(EmitSeparateAlwaysBlocks, bool);
-DECLARE_FIRTOOL_OPTION(EtcDisableInstanceExtraction, bool);
-DECLARE_FIRTOOL_OPTION(EtcDisableRegisterExtraction, bool);
-DECLARE_FIRTOOL_OPTION(EtcDisableModuleInlining, bool);
-DECLARE_FIRTOOL_OPTION(AddVivadoRAMAddressConflictSynthesisBugWorkaround, bool);
-DECLARE_FIRTOOL_OPTION(CkgModuleName, MlirStringRef);
-DECLARE_FIRTOOL_OPTION(CkgInputName, MlirStringRef);
-DECLARE_FIRTOOL_OPTION(CkgOutputName, MlirStringRef);
-DECLARE_FIRTOOL_OPTION(CkgEnableName, MlirStringRef);
-DECLARE_FIRTOOL_OPTION(CkgTestEnableName, MlirStringRef);
-DECLARE_FIRTOOL_OPTION(ExportModuleHierarchy, bool);
-DECLARE_FIRTOOL_OPTION(StripFirDebugInfo, bool);
-DECLARE_FIRTOOL_OPTION(StripDebugInfo, bool);
+#define DECLARE_FIRTOOL_OPTIONS_CREATE_DESTROY(opt)                            \
+  MLIR_CAPI_EXPORTED Firtool##opt##Options firtool##opt##OptionsCreateDefault( \
+      FirtoolGeneralOptions general);                                          \
+  MLIR_CAPI_EXPORTED void firtool##opt##OptionsDestroy(                        \
+      Firtool##opt##Options options);                                          \
+  MLIR_CAPI_EXPORTED void firtool##opt##OptionsSetGeneral(                     \
+      Firtool##opt##Options options, FirtoolGeneralOptions general);           \
+  MLIR_CAPI_EXPORTED FirtoolGeneralOptions firtool##opt##OptionsGetGeneral(    \
+      Firtool##opt##Options options);
+
+#define DECLARE_FIRTOOL_OPTION(opt, name, type)                                \
+  MLIR_CAPI_EXPORTED void firtool##opt##OptionsSet##name(                      \
+      Firtool##opt##Options options, type value);                              \
+  MLIR_CAPI_EXPORTED type firtool##opt##OptionsGet##name(                      \
+      Firtool##opt##Options options)
+
+// ========== General ==========
+
+MLIR_CAPI_EXPORTED FirtoolGeneralOptions firtoolGeneralOptionsCreateDefault();
+MLIR_CAPI_EXPORTED void
+firtoolGeneralOptionsDestroy(FirtoolGeneralOptions options);
+DECLARE_FIRTOOL_OPTION(General, DisableOptimization, bool);
+DECLARE_FIRTOOL_OPTION(General, ReplSeqMem, bool);
+DECLARE_FIRTOOL_OPTION(General, ReplSeqMemFile, MlirStringRef);
+DECLARE_FIRTOOL_OPTION(General, IgnoreReadEnableMem, bool);
+DECLARE_FIRTOOL_OPTION(General, IgnoreReadEnableMem, bool);
+DECLARE_FIRTOOL_OPTION(General, DisableRandom, FirtoolRandomKind);
```

Comment:
Other than the disabling optimization, I think these option should duplicate to decouple logic.

Reply:
but of cause, share the `FirtoolRandomKind` type

Reply:
These general options are used in multiple populate functions, and I personally prefer to keep the GeneralOptions so that there is only one storage for these options instead of multiple. It should be flexible enough for C-API users to choose whether they want to use the same shared GeneralOptions or separate GeneralOptions in all their other Options structures.

Reply:
IMHO, the duplication is inevitable... Although every one want to dedup.

For example, if adding more and more populate functions, e.g. 10+, and two functions have a same option, should we dedup them in to a `GeneralOption`? If so, for other populate functions, they got an unused option... Even for the `disableOpt` option, what if some functions don't use it? should it still remain in the options for these functions?

here is my 50 cents: for each C-API of populate function, it should have its own populate option, so that the library designer and following contributor can have more understanding to each populate functions.

But of cause, I think this decision should be made by upstream devs.

Reply:
FWIW, this is the difference between whether we use or don't use GeneralOptions.

With GeneralOptions (pseudo-code):
```cpp
struct CommandLineOpts {
    GeneralOpts generalOpts;
    CHIRRTLToLowFIRRTLOptions chirrtlToLowFIRRTLOpts{&generalOpts};
    HWToSVOptions hwToSVOpts{&generalOpts};

    cl::opts<bool, true> disableOptimization{cl::location(&generalOpts.disableOptimization)};
    cl::opts<bool, true> replSeqMem{cl::location(&generalOpts.replSeqMem)};
    cl::opts<std::string, true> replSeqMemFile{cl::location(&generalOpts.replSeqMemFile)};

    const auto &getCHIRRTLToLowFIRRTLOpts() const { return chirrtlToLowFIRRTLOpts; }
    const auto &getHWToSVOpts() const { return hwToSVOpts; }
};
```

Without GeneralOptions (pseudo-code):
```cpp
struct CommandLineOpts {
    CHIRRTLToLowFIRRTLOptions chirrtlToLowFIRRTLOpts;
    HWToSVOptions hwToSVOpts;

    mutable cl::opts<bool> disableOptimization;
    mutable cl::opts<bool> replSeqMem;
    mutable cl::opts<std::string> replSeqMemFile;

    const auto &getCHIRRTLToLowFIRRTLOpts() const {
        chirrtlToLowFIRRTLOpts.disableOptimization = this->disableOptimization.getValue();
        chirrtlToLowFIRRTLOpts.replSeqMem = this->replSeqMem.getValue();
        chirrtlToLowFIRRTLOpts.replSeqMemFile = this->replSeqMemFile.getValue();
        return firrtrlToHWOpts;
    }
    const auto &getHWToSVOpts() const {
        hwToSVOpts.disableOptimization = this->disableOptimization.getValue();
        hwToSVOpts.replSeqMem = this->replSeqMem.getValue();
        hwToSVOpts.replSeqMemFile = this->replSeqMemFile.getValue();
        return hwToSVOpts;
    }
};
```